File: typescript/src/patterns/fn/FunctionalStyleFactory.ts
/** Java: Supplier<T>. */
type Supplier<T> = () => T;

// Java: nested interface/classes of FunctionalStyleFactory
export interface Shape {
  draw(): void;
}

export class Circle implements Shape {
  draw(): void {
    console.log('Circle');
  }
}

export class Square implements Shape {
  draw(): void {
    console.log('Square');
  }
}

export function run(): void {
  const circleFactory: Supplier<Shape> = () => new Circle();
  const squareFactory: Supplier<Shape> = () => new Square(); // Java: Square::new
  const s1 = circleFactory();
  const s2 = squareFactory();
  s1.draw();
  s2.draw();
}

File: csharp/src/Training.Patterns/Fn/FunctionalStyleFactory.cs
namespace Training.Patterns.Fn;

public class FunctionalStyleFactory
{
    internal interface IShape
    {
        void Draw();
    }

    internal class Circle : IShape
    {
        public void Draw()
        {
            Console.WriteLine("Circle");
        }
    }

    internal class Square : IShape
    {
        public void Draw()
        {
            Console.WriteLine("Square");
        }
    }

    public static void Run()
    {
        Func<IShape> circleFactory = () => new Circle();
        Func<IShape> squareFactory = () => new Square(); // Java: Square::new
        var s1 = circleFactory();
        var s2 = squareFactory();
        s1.Draw();
        s2.Draw();
    }
}
